### PR TITLE
Button: Adding BaseButton, cleaning up css.

### PR DESCRIPTION
**Pull request by dzearing**
#### Pull request checklist

- [X] Include a change request file if publishing 

#### Description of changes

This refactors out Button into a BaseButton, which allows variants to subclass from. This ensures that we always implement `IButton`, while also minifying the amount of boilerplate you need to go through in order to create a variant.

#### Focus areas to test

Buttons should continue to work as is. I did a pass on the CSS to clean up messy things like asymmetrical paddings, wonky alignment hacks, and unnecessary rules.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Button/Button.tsx [hunk 1]
```diff
@@ -1,137 +1,48 @@
...
+import { BaseButton } from './BaseButton';
 
-  public render(): JSX.Element {
-    let { buttonType, children, icon, description, ariaLabel, ariaDescription, href, disabled, onClick } = this.props;
-    let { labelId, descriptionId, ariaDescriptionId } = this.state;
+import { DefaultButton } from './DefaultButton/DefaultButton';
+import { CommandButton } from './CommandButton/CommandButton';
+import { CompoundButton } from './CompoundButton/CompoundButton';
+import { IconButton } from './IconButton/IconButton';
+import { PrimaryButton } from './PrimaryButton/PrimaryButton';
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/Button.tsx]: Are you removing some of the css associated with this button? More importantly does the css and look of this button change at all? Just might be soemthing that could break it if people are adjusting for it.

File: packages/office-ui-fabric-react/src/components/Button/BaseButton.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      this.onRenderAriaDescription(),
+      this.onRenderChildren()
+    ];
+  }
+
+  protected onRenderIcon() {
+    let { icon } = this.props;
+
+    return icon ? (
+      <span className={ `${this._baseClassName}-icon` }>
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/BaseButton.tsx]: I know you can easily subclass this, but should you use the icon class instead? It allows more diversity with iconProps from the get-go.

File: packages/office-ui-fabric-react/src/components/Button/CompoundButton/CompoundButton.scss [hunk 3]
```diff
@@ -8,34 +8,38 @@
...
+  @include text-align(left);
 
   .ms-Button-label {
     @include ms-font-m;
-    @include text-align(left);
     display: block;
     font-weight: $ms-font-weight-semibold;
     color: $ms-color-black;
-    margin-top: -5px;
+    margin: 0 0 2px;
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/CompoundButton/CompoundButton.scss]: Needs to be rtl friendly.

File: packages/office-ui-fabric-react/src/components/Button/Button.tsx [hunk 4]
```diff
@@ -1,137 +1,48 @@
...
+    let props = this.props;
 
-    // ms-Button-description is only shown when the button type is compound.
-    // In other cases it will not be displayed.
-    const descriptionSpan: React.ReactElement<React.HTMLProps<HTMLSpanElement>> = description
-      ? <span className='ms-Button-description' id={ descriptionId }>{ description }</span>
-      : null;
+    switch (props.buttonType) {
+      case ButtonType.command:
+        return <CommandButton ref={ this._resolveRef('_button') } { ...props } />;
```
Comment (joschect) [inline packages/office-ui-fabric-react/src/components/Button/Button.tsx]: I think the order of these might need to be reversed. Unless you want someone to be able to override your resolveRef.

**Comment by micahgodbolt** [comment]
@dzearing  If all buttons now us the same set of props, shouldn't we remove CompoundButton.Props.ts and such?

File: packages/office-ui-fabric-react/src/components/Button/CompoundButton/CompoundButton.scss (hunk 3 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Button/CompoundButton/CompoundButton.scss]: nope. Top Right Bottom. Left and Right are both 0.

File: packages/office-ui-fabric-react/src/components/Button/BaseButton.tsx (hunk 2 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Button/BaseButton.tsx]: I would agree in general. I can look into it.

File: packages/office-ui-fabric-react/src/components/Button/Button.tsx (hunk 1 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Button/Button.tsx]: Perhaps; by removing the `&+&` classes I might be affecting scenarios where buttons show up next to each other. Dialog maybe. I will double check.

File: packages/office-ui-fabric-react/src/components/Button/Button.tsx (hunk 4 above)
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Button/Button.tsx]: i didn't think refs got passed through but i'll update anyway.

**Comment by dzearing** [comment]
I've removed the props.

**Comment by dzearing** [comment]
@micahgodbolt I think i don't want to update any of the docs in this change. Technically nothing should have changed in terms of usage.

**Comment by fantua** [comment]
@dzearing so, how to use hero button now?

**Pull request merged**